### PR TITLE
Add empty case class params check. Add test.

### PR DESCRIPTION
**Pull request by ashwinbhaskar**
Fixes https://github.com/lampepfl/dotty/issues/6969

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 1]
```diff
@@ -2396,6 +2396,11 @@ object Parsers {
         }
         else Nil
       }
+
+      if (ofCaseClass && in.token != LPAREN)
+        syntaxError( "case classes without a parameter list are not allowed;\n"+
+          "use either case objects or case classes with an explicit `()' as a parameter list.", in.lastOffset)
+
       recur(firstClause = true, 0, ofInstance)
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: I would do something along the lines of:

```scala
val start = in.offset
val paramss = recur(firstClause = true, 0, ofInstance)

if (ofCaseClass && paramss.isEmpty)
  syntaxError(..., start)
```

Comment (ashwinbhaskar) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: makes sense. Thank you :+1:

**Comment by ashwinbhaskar** [comment]
@allanrenucci can this be merged ?

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 2]
```diff
@@ -2396,7 +2396,15 @@ object Parsers {
         }
         else Nil
       }
-      recur(firstClause = true, 0, ofInstance)
+
+      val start = in.lastOffset
+      val params = recur(firstClause = true, 0, ofInstance)
+
+      if(ofCaseClass && params.isEmpty)
```
[the remark this answers is not in the document]
Comment (ashwinbhaskar) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: done

**Comment by ashwinbhaskar** [comment]
> We already have some broken logic in `Desugar.scala`:
> 
> https://github.com/lampepfl/dotty/blob/073c7aa47672c02932904dcd50bf5693d90d7299/compiler/src/dotty/tools/dotc/ast/Desugar.scala#L449-L451
> 
> We should unify both. Let's fix the code in Desugar. Then we probably don't need to report an error in the Parser

@allanrenucci Not exactly sure where `Desugar` comes into picture. I am guessing it comes before  `Parser`? If so isn't this code
```
if (isCaseClass && originalTparams.isEmpty) 
   ctx.error(CaseClassMissingParamList(cdef), namePos) 
```
taking care of reporting the `MissingParam` error?

**Comment by ashwinbhaskar** [comment]
@allanrenucci On a side note, is there a way to run only a particular test? 

```
sbt "test:testOnly *EmptyCaseClassParams.scala"
```
does not seem to work.

**Comment by allanrenucci** [comment]
> is there a way to run only a particular test?

Yes. Within sbt:
```
sbt:dotty> testCompilation EmptyCaseClassParams
```
Alternatively, you can compile a specific file using:
```
sbt:dotty> dotc path/to/file.scala
```

> Not exactly sure where Desugar comes into picture. I am guessing it comes before Parser? 

No it comes right after Parser

>  If so isn't this code ... taking care of reporting the MissingParam error?

Yes, but it currently does not report an error when the type parameter list in non empty. Removing `originalTparams.isEmpty` should fix the issue. Not sure if it will handle enums as well.

File: tests/neg/EmptyCaseClassParams.scala [hunk 3]
```diff
@@ -0,0 +1,3 @@
+object EmptyCaseClassParams{
+  case class A[T] // error
+}
```
[the remark this answers is not in the document]
Comment (ashwinbhaskar) [reply tests/neg/EmptyCaseClassParams.scala]: done

**Comment by ashwinbhaskar** [comment]
@allanrenucci done the changes. The Desugar file fix suffices, even for enums

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 4]
```diff
@@ -2396,6 +2396,7 @@ object Parsers {
         }
         else Nil
       }
+
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: Please remove spurious new line

Comment (ashwinbhaskar) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala]: done

**Comment by ashwinbhaskar** [comment]
> Otherwise LGTM. Thanks a lot @ashwinbhaskar!

more than happy to do it 😄

**Comment by nicolasstucki** [comment]
Thanks @ashwinbhaskar. Nice fix.

**Comment by nicolasstucki** [comment]
Just a small last detail. Could you change the commit message to `Fix #6969: Add empty case class params check` so that github will link the commit with the issue.

**Comment by ashwinbhaskar** [comment]
> Just a small last detail. Could you change the commit message to `Fix #6969: Add empty case class params check` so that github will link the commit with the issue.

Sure, done.

**Pull request merged**